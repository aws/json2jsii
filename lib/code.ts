export class Code {
  public readonly lines = new Array<string>();
  private indent = 0;

  public open(text: string = '') {
    this.line(text);
    this.indent += 2;
  }

  public close(text: string = '') {
    this.indent -= 2;
    this.line(text);
  }

  public openBlock(text: string) {
    this.open(`${text} {`);
  }

  public closeBlock() {
    this.close('}');
  }

  public line(text: string = ''): void {
    const line = ' '.repeat(this.indent) + text;
    this.lines.push(line);
  }

  public render(): string {
    return this.lines.join('\n');
  }
}