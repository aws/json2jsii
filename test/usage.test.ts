import { TypeGenerator } from '../lib';
import { CodeMaker } from 'codemaker';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

test('example with $ref', async () => {
  const g = new TypeGenerator({
    definitions: {
      Name: {
        description: 'Represents a name of a person',
        required: [ 'firstName', 'lastName' ],
        properties: {
          firstName: {
            type: 'string',
            description: 'The first name of the person',
          },
          lastName: {
            type: 'string',
            description: 'The last name of the person',
          },
        },
      },
    },
  });

  g.emitType('Person', {
    required: [ 'name' ],
    properties: {
      name: {
        description: 'The person\'s name',
        $ref: '#/definitions/Name',
      },
      color: {
        description: 'Favorite color. Default is green',
        enum: [ 'red', 'green', 'blue', 'yellow' ],
      },
    },
  });

  const code = new CodeMaker();
  code.openFile('person.ts');
  g.writeToCodeMaker(code);
  code.closeFile('person.ts');

  const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'json2jsii'));
  await code.save(outdir);

  expect(fs.readFileSync(path.join(outdir, 'person.ts'), 'utf-8')).toMatchSnapshot();
});

test('fails when trying to resolve an undefined $ref', () => {
  const g = new TypeGenerator();
  expect(() => g.emitType('Foo', { $ref: 'unresolvable' })).toThrow(/invalid \$ref {\"\$ref\":\"unresolvable\"}/);
  expect(() => g.emitType('Foo', { $ref: '#/definitions/unresolvable' })).toThrow(/unable to find a definition for the \$ref \"unresolvable\"/);
});