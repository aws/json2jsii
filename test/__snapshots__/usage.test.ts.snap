// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example with $ref 1`] = `
"/**
 * @schema Person
 */
export interface Person {
  /**
   * The person's name
   *
   * @schema Person#name
   */
  readonly name: Name;

  /**
   * Favorite color. Default is green
   *
   * @default green
   * @schema Person#color
   */
  readonly color?: PersonColor;

}

/**
 * Converts an object of type 'Person' to JSON representation.
 */
export function Person_toJson(obj: Person | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'name': Name_toJson(obj.name),
    'color': obj.color,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}

/**
 * Represents a name of a person
 *
 * @schema Name
 */
export interface Name {
  /**
   * The first name of the person
   *
   * @schema Name#firstName
   */
  readonly firstName: string;

  /**
   * The last name of the person
   *
   * @schema Name#lastName
   */
  readonly lastName: string;

}

/**
 * Converts an object of type 'Name' to JSON representation.
 */
export function Name_toJson(obj: Name | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'firstName': obj.firstName,
    'lastName': obj.lastName,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}

/**
 * Favorite color. Default is green
 *
 * @default green
 * @schema PersonColor
 */
export enum PersonColor {
  /** red */
  RED = \\"red\\",
  /** green */
  GREEN = \\"green\\",
  /** blue */
  BLUE = \\"blue\\",
  /** yellow */
  YELLOW = \\"yellow\\",
}
"
`;
